### PR TITLE
Nanopore Fastq Host Removal Pipeline

### PR DESCRIPTION
**Pull request by DarianHole**
## Summary:
Adding a pipeline to do nanopore fastq host removal

## Changes:
- Wrote new host removal process based on minimap2
  - Fast5 host removal available
  - Folder structure of the output matches nanostripper
- Variable name changes and additions
  - More to come in changelog and after adjustments to the processes here as names and configs are a bit messy
- pysam added to nanopore environment
- Added a help command
- Adjusted resources for each process along with new tags and labels
- Update CHANGELOG
- Update README

## Potential Changes:
- Can maybe switch to 1 dehost script and have functions for nanopore and illumina inputs

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -5,17 +5,39 @@ Nextflow pipeline that removes human reads from SARS-CoV-2 Illumina and Nanopore
...
+**Nanopore Minimap2** - Competitive mapping approach based around removing human reads from either input fastq files or barcoded directories while maintaining as many viral reads as possible. Optional fast5 dehosting available in this pipeline
 
-*Currently* nanopore dehosting is still a little rough and you will also have to provide a path to the nanostripper environment along with the tool itself with the following:
+**Nanopore Nanostripper** - Dual mapping approach based around [nanostripper](https://github.com/nodrogluap/nanostripper) and then subsequently [guppy](https://nanoporetech.com/nanopore-sequencing-data-analysis) to generate de-hosted, demultiplexed fast5 and fastq files from input fast5 files. 
+
+*Currently* the nanopore nanostripper dehosting pipeline is still rough to run and probably will be for a while. You will have to provide a path to the guppy environment(s), the nanostripper environment and the nanostripper tool itself with the following arguments to generate a completely dehosted run:
```
Comment (ConnorChato) [inline README.md]: Could use some time estimate here. And make this a sub-heading under the Nanopore Nanostripper list item just for visuals.

File: README.md [hunk 2]
```diff
@@ -170,23 +212,217 @@ Found in `./results/` directory, the outputs for the Illumina pipeline include:
...
+    - Inputs:
+        - Dehosted bam file
+
+    - Outputs:
+        - Dehosted fastq file
+    
+    - Tools:
+        - samtools fastq
+
+6. (OPTIONAL) Regenerate fast5 files and the sequencing summary file
```
Comment (ConnorChato) [inline README.md]: Is this connected to a specific optional flag above?

File: bin/dehost_nanopore.py [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+    header = bamfile.header
+    # Removal of host reads if they are above threshold
+    for read in bamfile.fetch():
+        if read.reference_name != contig_ID and read.mapping_quality >= remove_minimum_quality:
+            reads_to_remove_set.add(read.query_name)
+            h_count += 1
+
+    # Have to go through again as we need to remove reads that have mapped twice, can probably do it in one pass but blanking on a good solution
+    for read in bamfile.fetch():
+        # Already found human reads, don't add them
```
Comment (ConnorChato) [inline bin/dehost_nanopore.py]: I think that this should work if you want to cut the pass in 30-33. 
```
if (read.reference_name != contig_ID and read.mapping_quality >= remove_minimum_quality) or read.query_name in reads_to_remove_set:
    h_count += 1
    pass
```
If you never plan on passing an actual value to reads_to_remove_set, you can also just cut that as well and do
```
if read.reference_name != contig_ID and read.mapping_quality >= remove_minimum_quality:
    h_count += 1
    pass
```

File: environments/nanopore.yml [hunk 4]
```diff
@@ -6,4 +6,7 @@ dependencies:
   - samtools
   - minimap2
   - ont-fast5-api
-  - csvtk
\ No newline at end of file
+  - csvtk
+  - pip
+  - pip:
+    - pysam
```
Comment (ConnorChato) [inline environments/nanopore.yml]: Unless there are specific config-y reasons, it is probably a good idea to end this with a newline

File: modules/help.nf [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+            --fastq_directory [path]            Path to directory containing either barcoded directories OR individual fastq files
+
+        Optional:
+            --fast5_directory [path]            Path to fast5 directories associated with the data
+            --composite_minimap2_index [file]   Path to composite minimap2 .mmi file to speed up analysis
+            --min_length [i]                    Minimum length of fastq reads to keep (Default: 400)
+            --max_length [i]                    Maximum length of fastq reads to keep (Default: 2400)
+
+        Example Command:
+            nextflow run phac-nml/ncov-dehoster --human_ref ./hg38.fa -profile conda --nanopore --minimap2 --fastq_directory ./fastqs
```
Comment (ConnorChato) [inline modules/help.nf:68]: just for the sake of example consistency, you could adjust hg38.fa to be .fasta as in line 7 here

**Review by ConnorChato — APPROVED** [review]
May not have too many suggestions for the nextflow code, but in general the obvious stuff checks out.

File: README.md (hunk 1 above)
Comment (DarianHole) [reply README.md]: Hopefully I've organized it a bit better now. I also added in the time it takes for this one but I should probably add the others? Hard to estimate them as its based on size + number of fastq files

File: README.md (hunk 2 above)
Comment (DarianHole) [reply README.md]: Yep its an optional flag, I've added it here so someone reading through will know how to run the optional step if they wanted to

File: modules/help.nf (hunk 5 above)
Comment (DarianHole) [reply modules/help.nf:68]: Changed all to .fa

File: bin/dehost_nanopore.py (hunk 3 above)
Comment (DarianHole) [reply bin/dehost_nanopore.py]: #15 Should be addressed now

**Pull request merged**